### PR TITLE
Make static context a feature and allow dynamic context generation

### PR DESCRIPTION
**Pull request by sorpaas**
This changes the static contexts to be a feature that can be enabled or disabled, and allow contexts to be dynamically generated. By disabling static contexts, users accomplish much smaller binary size, but the trade-off is then that contexts must be generated dynamically (which is not a cheap operation).

### CODE REVIEW

**Comment by dvdplm** [comment]
Can you ELI5 what "static|dynamic context" means in this context?

**Comment by sorpaas** [comment]
@dvdplm It's the pre-computed table used to accelerate various calculations. Previously we only allow storing the table as static variables directly in the library. This PR allows disabling those variables (to reduce binary size, if needed) and generate the tables dynamically.

File: core/src/ecmult.rs [hunk 1]
```diff
@@ -9,18 +13,149 @@ pub const ECMULT_TABLE_SIZE_A: usize = 1 << (WINDOW_A - 2);
...
+        }
+    }
+    ret
+}
+
+/// Calculate the (modular) inverses of a batch of field
+/// elements. Requires the inputs' magnitudes to be at most 8. The
+/// output magnitudes are 1 (but not guaranteed to be
+/// normalized). The inputs and outputs must not overlap in
+/// memory.
```
Comment (ordian) [inline core/src/ecmult.rs]: what do you mean by "The inputs and outputs must not overlap in memory"? outputs are not supplied by a user here?

File: core/src/ecmult.rs [hunk 2]
```diff
@@ -9,18 +13,149 @@ pub const ECMULT_TABLE_SIZE_A: usize = 1 << (WINDOW_A - 2);
...
+pub fn inv_all_var(fields: &[Field]) -> Vec<Field> {
+    if fields.len() == 0 {
+        return Vec::new();
+    }
+
+    let mut ret = Vec::new();
```
Comment (ordian) [inline core/src/ecmult.rs]:
```suggestion
    let mut ret = Vec::with_capacity(fields.len());
```

File: core/src/ecmult.rs [hunk 3]
```diff
@@ -9,18 +13,149 @@ pub const ECMULT_TABLE_SIZE_A: usize = 1 << (WINDOW_A - 2);
...
 /// Context for accelerating the computation of a*P + b*G.
 pub struct ECMultContext {
     pre_g: [AffineStorage; ECMULT_TABLE_SIZE_G],
 }
 
 impl ECMultContext {
-    /// Create a new `ECMultContext`.
-    pub const fn new(pre_g: [AffineStorage; ECMULT_TABLE_SIZE_G]) -> Self {
+    /// Create a new `ECMultContext` from raw values.
+    pub const unsafe fn new_raw(pre_g: [AffineStorage; ECMULT_TABLE_SIZE_G]) -> Self {
```
Comment (ordian) [inline core/src/ecmult.rs]: what makes it unsafe? is it due to potential misuse in crypto logic and not in rust terms?

File: core/src/ecmult.rs [hunk 4]
```diff
@@ -9,18 +13,149 @@ pub const ECMULT_TABLE_SIZE_A: usize = 1 << (WINDOW_A - 2);
...
+pub fn set_all_gej_var(a: &[Jacobian]) -> Vec<Affine> {
+    let mut az: Vec<Field> = Vec::with_capacity(a.len());
+    for i in 0..a.len() {
+        if !a[i].is_infinity() {
+            az.push(a[i].z.clone());
+        }
+    }
+    let azi: Vec<Field> = inv_all_var(&az);
+
+    let mut ret = Vec::with_capacity(a.len());
```
Comment (dvdplm) [inline core/src/ecmult.rs]: Couldn't this be? (`Affine` needs to be `Clone` I guess?)
```
let ret = vec![Affine::default(); a.len()];`
```

File: core/src/ecmult.rs [hunk 5]
```diff
@@ -9,18 +13,149 @@ pub const ECMULT_TABLE_SIZE_A: usize = 1 << (WINDOW_A - 2);
...
+    }
+
+    /// Deconstruct the `ECMultContext` into raw `AffineStorage`.
+    pub fn deconstruct_raw(self) -> [AffineStorage; ECMULT_TABLE_SIZE_G] {
+        self.pre_g
+    }
+
+    /// Generate a new `ECMultContext` in the heap. Note that this function is expensive.
+    pub fn new_boxed() -> Box<Self> {
+        let mut this = unsafe {
```
Comment (ordian) [inline core/src/ecmult.rs]: A comment justifying the usage of `unsafe` won't hurt future maintainers :)

File: core/src/ecmult.rs [hunk 6]
```diff
@@ -9,18 +13,149 @@ pub const ECMULT_TABLE_SIZE_A: usize = 1 << (WINDOW_A - 2);
...
+        let i = i - 1;
+        ret[j] = &ret[i] * &u;
+        u = &u * &fields[j];
+    }
+
+    ret[0] = u;
+    ret
+}
+
+const GEN_BLIND: Scalar = Scalar(
```
Comment (dvdplm) [inline core/src/ecmult.rs]: For the un-initiated reader it would be useful to have links to where these numbers come from (if such a canonical source exists; a comment explaining how to generate them if not)

File: core/src/ecmult.rs [hunk 7]
```diff
@@ -9,18 +13,149 @@ pub const ECMULT_TABLE_SIZE_A: usize = 1 << (WINDOW_A - 2);
...
+    ),
+    z: Field::new_raw(
+        3685368, 75404844, 20246216, 5748944, 73206666,
+        107661790, 110806176, 73488774, 5707384, 104448710
+    ),
+    infinity: false,
+};
+
 /// Context for accelerating the computation of a*G.
 pub struct ECMultGenContext {
```
Comment (dvdplm) [inline core/src/ecmult.rs]: Maybe dumb q but does this have to be `pub`?

File: core/src/ecmult.rs [hunk 8]
```diff
@@ -29,13 +164,87 @@ pub struct ECMultGenContext {
...
 impl ECMultGenContext {
-    /// Create a new `ECMultGenContext`.
-    pub const fn new(
-        prec: [[AffineStorage; 16]; 64],
-        blind: Scalar,
-        initial: Jacobian
-    ) -> Self {
-        Self { prec, blind, initial }
+    /// Create a new `ECMultGenContext` from raw values.
+    pub const unsafe fn new_raw(prec: [[AffineStorage; 16]; 64]) -> Self {
```
Comment (dvdplm) [inline core/src/ecmult.rs]: Document safety

File: core/src/ecmult.rs [hunk 9]
```diff
@@ -29,13 +164,87 @@ pub struct ECMultGenContext {
...
+    pub fn inspect_raw(&self) -> &[[AffineStorage; 16]; 64] {
+        &self.prec
+    }
+
+    /// Deconstruct `ECMultGenContext` into raw values.
+    pub fn deconstruct_raw(self) -> [[AffineStorage; 16]; 64] {
+        self.prec
+    }
+
+    /// Generate a new `ECMultGenContext` in the heap. Note that this function is expensive.
```
Comment (dvdplm) [inline core/src/ecmult.rs]:
```suggestion
    /// Generate a new `ECMultGenContext` on the heap. Note that this function is expensive.
```

File: core/src/ecmult.rs [hunk 10]
```diff
@@ -29,13 +164,87 @@ pub struct ECMultGenContext {
...
+    pub fn new_boxed() -> Box<Self> {
+        let mut this = unsafe {
```
Comment (dvdplm) [inline core/src/ecmult.rs]: Document safety

File: src/lib.rs [hunk 11]
```diff
@@ -1,6 +1,19 @@
+//! Pure Rust implementation of the secp256k1 curve and fast ECDSA
+//! signatures. The secp256k1 curve is used excusively in Bitcoin and
```
Comment (dvdplm) [inline src/lib.rs]:
```suggestion
//! signatures. The secp256k1 curve is used extensively in Bitcoin and
```

File: src/lib.rs [hunk 12]
```diff
@@ -1,6 +1,19 @@
...
+//! Ethereum alike cryptocurrencies.
```
Comment (dvdplm) [inline src/lib.rs]:
```suggestion
//! Ethereum-alike cryptocurrencies.
```

File: core/src/ecmult.rs (hunk 1 above)
Comment (sorpaas) [reply core/src/ecmult.rs]: I think that was from C library and does not really apply here. Let me remove it.

File: core/src/ecmult.rs (hunk 3 above)
Comment (sorpaas) [reply core/src/ecmult.rs]: If the table is not filled in the correct data of `pre_g`, then usage of it would be invalid. So yeah it's in crypto logic, but also not sure if it will lead to anything panicking.

File: core/src/ecmult.rs [hunk 13]
```diff
@@ -9,18 +13,149 @@ pub const ECMULT_TABLE_SIZE_A: usize = 1 << (WINDOW_A - 2);
...
         Self { pre_g }
     }
+
+    /// Inspect raw values of `ECMultContext`.
+    pub fn inspect_raw(&self) -> &[AffineStorage; ECMULT_TABLE_SIZE_G] {
+        &self.pre_g
+    }
+
+    /// Deconstruct the `ECMultContext` into raw `AffineStorage`.
+    pub fn deconstruct_raw(self) -> [AffineStorage; ECMULT_TABLE_SIZE_G] {
```
[the remark this answers is not in the document]
Comment (sorpaas) [reply core/src/ecmult.rs]: Let me just remove it. I figured out we actually don't need this function!

File: core/src/ecmult.rs (hunk 6 above)
Comment (sorpaas) [reply core/src/ecmult.rs]: Created an issue #43. Let me do this documentation work in a separate PR.

File: core/src/ecmult.rs (hunk 7 above)
Comment (sorpaas) [reply core/src/ecmult.rs]: Library users need it when they use dynamic context. Something like this:

```
let context: Box<ECMultGenContext> = ECMultGenContext::new_boxed();
let public = PublicKey::from_secret_key_with_context(&secret_key, &context);
```

**Comment by dvdplm** [comment]
> This PR allows disabling those variables (to reduce binary size, if needed) and generate the tables dynamically.

How much bigger is the static vs dynamic build? How much does the table generation cost under normal use?

I think it'd be good to get some of this info into the README to help users decide. :)

**Comment by sorpaas** [comment]
> How much bigger is the static vs dynamic build? How much does the table generation cost under normal use?

It is reported to be quite significant -- around 1mb or so. I knew some users were discouraged previously from using `libsecp256k1` just because of this. I don't have the exact numbers in hand but we can get it later. :)

**Comment by dvdplm** [comment]
> we can get it later

Added https://github.com/paritytech/libsecp256k1/issues/46
